### DISCUSSION TITLE
Are there any plans to create a new task for calculating the similarity of objects?

### DISCUSSION THREAD

**Discussion by Hezonar** [ultralytics/ultralytics#11807] [Q&A]
Maybe not a task, but just a function, but that would be great. 

The reason I’m asking you specifically, and not saying that a person should do this outside the library, is that this task also uses the approach of looking at the last layer of the neural network to see how different its outputs are.

Perhaps in architectural terms, implementing this task is not such a simple solution, but this task is also one of the primary tasks in the field of computer vision. It can be applied, for example, in finding similarity among faces or various objects. This could be useful, for instance, when using two or more cameras to associate objects with each other. I believe many more examples can be found where this task can be useful.

_P.s.: I'm very pleased with what you're doing and how rapidly the library is advancing. Most likely, it has become the most pivotal library in the field of computer vision at this point. I hope it continues to improve. All the best._

**Comment by pderrenger** [comment]
Thank you for your kind words and for sharing your idea about a new task for calculating the similarity of objects using neural network outputs!

At the moment, we don't have a dedicated task or function specifically for calculating object similarity built into YOLOv8. However, it's an interesting suggestion that aligns well with advanced computer vision applications, such as facial recognition or multi-camera object tracking.

You can currently utilize features from the last layers of our models for similar purposes. By extracting embeddings from these layers, you can compute similarities using cosine similarity or Euclidean distance, for example.

Here’s a basic outline of how you might approach this using features from a trained YOLO model:

```python
from ultralytics import YOLO
import torch.nn.functional as F

# Load the model
model = YOLO('yolov8n.pt')

# Feature extraction from the last layer
features1 = model('image1.jpg').embeddings
features2 = model('image2.jpg').embeddings

# Calculate cosine similarity
similarity = F.cosine_similarity(features1, features2)
print('Similarity:', similarity.item())
```

Your suggestion is indeed valuable, and I’ll forward it to the development team for consideration in future releases. Contributions and ideas from the community are always welcome and play a significant role in enhancing the usability and functionality of our tools.

**Reply by KDeser** [reply]
This results in an error message:

`AttributeError: 'list' object has no attribute 'embeddings'`

**Reply by pderrenger** [reply]
Hi there!

Thank you for bringing this to our attention. It looks like there might be a misunderstanding or an issue with the code snippet provided. Let's address this step-by-step:

1. **Reproducible Example**: Could you please provide a minimum reproducible code example? This will help us understand the context and reproduce the issue on our end. You can refer to our guide on creating a minimum reproducible example [here](https://docs.ultralytics.com/help/minimum_reproducible_example). This is crucial for us to investigate and resolve the issue effectively.

2. **Version Check**: Ensure that you are using the latest versions of `torch` and `ultralytics`. You can upgrade your packages using the following commands:
    ```bash
    pip install --upgrade torch ultralytics
    ```

3. **Code Correction**: The error message indicates that the `results` object is a list, and lists do not have an `embeddings` attribute. You should access the individual result objects within the list. Here’s an updated example:
    ```python
    from ultralytics import YOLO
    import torch.nn.functional as F

    # Load the model
    model = YOLO('yolov8n.pt')

    # Run inference and get results
    results1 = model('image1.jpg')
    results2 = model('image2.jpg')

    # Assuming the embeddings are stored in the first result object
    features1 = results1[0].embeddings
    features2 = results2[0].embeddings

    # Calculate cosine similarity
    similarity = F.cosine_similarity(features1, features2)
    print('Similarity:', similarity.item())
    ```

If the issue persists after trying the above steps, please share the specific code you are using and any additional error messages you encounter. This will help us assist you more effectively.

Looking forward to your response! 😊
